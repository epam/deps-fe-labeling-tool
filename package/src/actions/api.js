import { createAction } from 'redux-actions'
import { rotationAnglesSelector } from '@/selectors/canvas'
import { pageImageUrlSelector } from '@/selectors/document'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { primaryLanguageSelector } from '@/selectors/ocr'
import { getApi } from '@/services/api'

const FEATURE_NAME = 'API'

const requestAttempt = createAction(
  `${FEATURE_NAME}/REQUEST_ATTEMPT`
)

const requestSuccess = createAction(
  `${FEATURE_NAME}/REQUEST_SUCCESS`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const requestFailure = createAction(
  `${FEATURE_NAME}/REQUEST_FAILURE`,
  (requestId, error) => ({
    requestId,
    error
  })
)

export const createRequestAction = (requestId, actionCreator) => {
  const requestActionCreator = (...args) => async (dispatch, getState) => {
    try {
      dispatch(requestAttempt(requestId))
      const result = await actionCreator(...args)(dispatch, getState)
      dispatch(requestSuccess(requestId))
      return result
    } catch (error) {
      dispatch(requestFailure(requestId, error.message))
      throw error
    }
  }
  requestActionCreator.toString = () => requestId
  return requestActionCreator
}

const save = createRequestAction(
  'save',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().save(markup, rotationAngles, language)
  }
)

const saveMarkup = createRequestAction(
  'saveMarkup',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    const fields = fieldsSelector(state)
    return getApi().saveMarkup(markup, rotationAngles, language, fields)
  }
)

const ocrText = createRequestAction(
  'ocrText',
  (engine, coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrText(
      engine,
      pageImageUrl,
      coordinates,
      language
    )
  }
)

const omrArea = createRequestAction(
  'omrArea',
  (coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)

    return getApi().omrArea(
      pageImageUrl,
      coordinates
    )
  }
)

const detectTables = createRequestAction(
  'detectTables',
  (selectedAreaCoords) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    return getApi().detectTables(
      pageImageUrl,
      selectedAreaCoords
    )
  }
)

const ocrTable = createRequestAction(
  'ocrTable',
  (engine, table) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrTable(
      engine,
      pageImageUrl,
      table,
      language
    )
  }
)

export {
  detectTables,
  ocrTable,
  ocrText,
  omrArea,
  resetDefault,
  requestAttempt,
  requestSuccess,
  requestFailure,
  save,
  saveMarkup
}
