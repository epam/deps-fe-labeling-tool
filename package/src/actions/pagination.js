import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
