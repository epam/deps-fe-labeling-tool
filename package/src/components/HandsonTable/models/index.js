
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
