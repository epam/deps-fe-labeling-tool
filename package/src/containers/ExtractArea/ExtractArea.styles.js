import styled from 'styled-components'
import { Button } from '@/components/Button'
import { Divider } from '@/components/Divider'

const StyledDivider = styled(Divider)`
  margin: 0 !important;
`

const StyledButton = styled(Button)`
  font-size: 1.4rem;
  line-height: 2.2rem;
  font-weight: 600;
`

export {
  StyledDivider as Divider,
  StyledButton as Button
}
