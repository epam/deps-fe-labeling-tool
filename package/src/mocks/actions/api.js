import { mockMarkupSelectors } from '@/mocks/selectors/markup'

const omrArea = jest.fn(() => Promise.resolve({
  content: false,
  confidence: 0.77
}))

const ocrTable = jest.fn(() => Promise.resolve({ type: 'MOCK_OCR_TABLE' }))

const ocrText = jest.fn(() => Promise.resolve({
  content: 'MOCK_OCR_TEXT',
  confidence: 0.98
}))

const save = jest.fn(() => Promise.resolve(mockMarkupSelectors.markupSelector()))

const saveMarkup = jest.fn(() => Promise.resolve({ type: 'MOCK_SAVE_MARKUP' }))

const detectTables = jest.fn((url, coords) => Promise.resolve(['markupTable']))

const mockApiActions = {
  detectTables,
  omrArea,
  ocrTable,
  ocrText,
  save,
  saveMarkup
}

export {
  mockApiActions
}
