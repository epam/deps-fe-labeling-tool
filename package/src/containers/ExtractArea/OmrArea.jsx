import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { omrArea } from '@/actions/api'
import { updateLabels } from '@/actions/markup'
import { ButtonType } from '@/components/Button'
import { ObjectTitle } from '@/components/ObjectTitle'
import {
  Label,
  labelShape
} from '@/models/Label'
import { currentPageSelector } from '@/selectors/pagination'
import { getApi } from '@/services/api'
import {
  Divider,
  Button
} from './ExtractArea.styles'

const RESOURCE_OMR = 'OMR'
const RESOURCE_EXTRACT_DATA = 'Extract data'

const OmrArea = ({
  currentPage,
  label,
  omrArea,
  updateLabels
}) => {
  const omr = async () => {
    const savedLabel = label
    const coords = Label.toRectangle(savedLabel)
    const { content, confidence } = await omrArea(coords)
    updateLabels(
      currentPage,
      [{
        ...savedLabel,
        content,
        confidence
      }]
    )
  }

  const isOMRAvailable = !!getApi().omrArea

  if (!isOMRAvailable) {
    return null
  }

  return (
    <>
      <Divider />
      <ObjectTitle
        title={RESOURCE_OMR}
      />
      <Button
        onClick={omr}
        type={ButtonType.PRIMARY}
      >
        {RESOURCE_EXTRACT_DATA}
      </Button>
    </>
  )
}

OmrArea.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  omrArea: PropTypes.func.isRequired,
  updateLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state)
})

const mapDispatchToProps = {
  omrArea,
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(OmrArea)

export {
  ConnectedComponent as OmrArea
}
