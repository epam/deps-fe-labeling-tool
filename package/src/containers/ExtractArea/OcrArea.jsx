import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { ocrText } from '@/actions/api'
import { updateLabels } from '@/actions/markup'
import { ButtonType } from '@/components/Button'
import { ObjectTitle } from '@/components/ObjectTitle'
import { OcrSelect } from '@/containers/OcrSelect'
import {
  Label,
  labelShape
} from '@/models/Label'
import { ocrEngineShape } from '@/models/OcrEngine'
import {
  primaryEngineSelector,
  ocrEnginesSelector
} from '@/selectors/ocr'
import { currentPageSelector } from '@/selectors/pagination'
import { getApi } from '@/services/api'
import {
  Divider,
  Button
} from './ExtractArea.styles'

const RESOURCE_OCR = 'OCR'
const RESOURCE_EXTRACT_DATA = 'Extract data'

const OcrArea = ({
  currentPage,
  label,
  ocrEngines,
  ocrText,
  primaryEngine,
  updateLabels
}) => {
  const ocr = async () => {
    const savedLabel = label
    const coords = Label.toRectangle(savedLabel)
    const { content, confidence } = await ocrText(primaryEngine, coords)
    updateLabels(
      currentPage,
      [{
        ...savedLabel,
        content,
        confidence
      }]
    )
  }

  const isOCRAvailable = !!ocrEngines?.length && !!getApi().ocrText

  if (!isOCRAvailable) {
    return null
  }

  return (
    <>
      <Divider />
      <ObjectTitle
        title={RESOURCE_OCR}
      />
      <OcrSelect
        ocr={ocr}
      />
      <Button
        onClick={ocr}
        type={ButtonType.PRIMARY}
      >
        {RESOURCE_EXTRACT_DATA}
      </Button>
    </>
  )
}

OcrArea.propTypes = {
  currentPage: PropTypes.number.isRequired,
  label: labelShape.isRequired,
  ocrEngines: PropTypes.arrayOf(ocrEngineShape),
  primaryEngine: PropTypes.string.isRequired,
  ocrText: PropTypes.func.isRequired,
  updateLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  ocrEngines: ocrEnginesSelector(state),
  primaryEngine: primaryEngineSelector(state)
})

const mapDispatchToProps = {
  ocrText,
  updateLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(OcrArea)

export {
  ConnectedComponent as OcrArea
}
