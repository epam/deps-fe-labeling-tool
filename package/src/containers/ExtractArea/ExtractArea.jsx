import React from 'react'
import { OcrArea } from '@/containers/ExtractArea/OcrArea'
import { OmrArea } from '@/containers/ExtractArea/OmrArea'
import { labelShape, LabelType } from '@/models/Label'

const ExtractArea = ({ label }) => {
  if (label.type === LabelType.CHECKMARK) {
    return <OmrArea label={label} />
  }

  return <OcrArea label={label} />
}

ExtractArea.propTypes = {
  label: labelShape.isRequired
}

export { ExtractArea }
