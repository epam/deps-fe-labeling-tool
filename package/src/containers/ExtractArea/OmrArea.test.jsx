import { mockApiActions } from '@/mocks/actions/api'
import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import React from 'react'
import { shallow } from 'enzyme'
import { omrArea } from '@/actions/api'
import { updateLabels } from '@/actions/markup'
import { Label, LabelType } from '@/models/Label'
import { getApi } from '@/services/api'
import { Button } from './ExtractArea.styles'
import { OmrArea } from './OmrArea'

jest.mock('@/services/api', () => ({
  getApi: jest.fn(() => ({
    omrArea: jest.fn(() => Promise.resolve({
      confidence: 0.11,
      content: false
    }))
  }))
}))
jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/actions/api', () => mockApiActions)

const {
  WrappedComponent,
  mapStateToProps,
  mapDispatchToProps
} = OmrArea

describe('Container: OmrArea', () => {
  describe('mapStateToProps', () => {
    it('should pass expected state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      expect(props).toMatchSnapshot()
    })
  })

  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should pass updateLabels action as prop to WrappedComponent', () => {
      props.updateLabels()
      expect(updateLabels).toHaveBeenCalled()
    })

    it('should pass omrArea action as prop to WrappedComponent', () => {
      props.omrArea()
      expect(omrArea).toHaveBeenCalled()
    })
  })

  describe('component', () => {
    let wrapper, defaultProps
    const testLabel = new Label(1, 2, 3, 4, undefined, 5, LabelType.STRING, 'test', {}, 0.78)

    beforeEach(() => {
      defaultProps = {
        ...mapStateToProps().props,
        ...mapDispatchToProps().props,
        label: testLabel
      }
      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render layout correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call updateLabels prop with correct args when calling Button onClick prop', async () => {
      const btn = wrapper.find(Button)
      await btn.props().onClick()
      const expectedSecondArg = [
        {
          ...defaultProps.label,
          ...await defaultProps.omrArea()
        }
      ]
      expect(defaultProps.updateLabels).nthCalledWith(1, defaultProps.currentPage, expectedSecondArg)
    })

    it('should not render anything in case api of omrArea is not provided', () => {
      getApi.mockImplementationOnce(() => () => ({
        omrArea: undefined
      }))
      const wrapper = shallow(<WrappedComponent {...defaultProps} />)
      expect(wrapper.isEmptyRender()).toEqual(true)
    })
  })
})
