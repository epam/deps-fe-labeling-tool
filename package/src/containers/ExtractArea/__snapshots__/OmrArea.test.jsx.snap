// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: OmrArea component should render layout correctly 1`] = `
<Fragment>
  <ExtractAreastyles__StyledDivider />
  <ObjectTitle
    title="OMR"
  />
  <ExtractAreastyles__StyledButton
    onClick={[Function]}
    type="primary"
  >
    Extract data
  </ExtractAreastyles__StyledButton>
</Fragment>
`;

exports[`Container: OmrArea mapStateToProps should pass expected state as props to WrappedComponent 1`] = `
Object {
  "currentPage": 1,
}
`;
